dataframely/schema.py

```diff
@@ -876,6 +879,68 @@ def _requires_validation_for_reading_parquet(
             )
         return True
 
+    # ------------------------------------- IO --------------------------------------- #
+    @classmethod
+    def _write(cls, df: pl.DataFrame, io: StorageBackend, **kwargs: Any) -> None:
+        io.write_frame(df=df, serialized_schema=cls.serialize(), **kwargs)
+
+    @classmethod
+    def _sink(cls, lf: pl.LazyFrame, io: StorageBackend, **kwargs: Any) -> None:
+        io.sink_frame(lf=lf, serialized_schema=cls.serialize(), **kwargs)
+
+    @overload
+    @classmethod
+    def _read(
+        cls,
+        io: StorageBackend,
+        validation: Validation,
+        lazy: Literal[True],
+        **kwargs: Any,
+    ) -> LazyFrame[Self]: ...
+
+    @overload
+    @classmethod
+    def _read(
+        cls,
+        io: StorageBackend,
+        validation: Validation,
+        lazy: Literal[False],
+        **kwargs: Any,
+    ) -> DataFrame[Self]: ...
+
+    @classmethod
+    def _read(
+        cls, io: StorageBackend, validation: Validation, lazy: bool, **kwargs: Any
+    ) -> LazyFrame[Self] | DataFrame[Self]:
+        source = kwargs.pop("source")
+
+        # Load
+        if lazy:
+            lf, serialized_schema = io.scan_frame(source=source)
+        else:
+            df, serialized_schema = io.read_frame(source=source)
+            lf = df.lazy()
+
+        deserialized_schema = (
+            deserialize_schema(serialized_schema) if serialized_schema else None
+        )
+
+        # Smart validation
+        if cls._requires_validation_for_reading_parquet(
+            deserialized_schema, validation, source=str(source)
+        ):
+            validated = cls.validate(lf, cast=True)
+            if lazy:
+                return validated.lazy()
+            else:
+                return validated
+
+        casted = cls.cast(lf)
```

Comment:
Did we always cast, no matter what? 🤔

Reply:
If validation is required, we already return inside the if-block right above this. So we only cast if validation is not required, which seems sensible to me.